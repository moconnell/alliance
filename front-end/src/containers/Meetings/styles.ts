import styled from "styled-components";

export const Container = styled.div`
  height: 100%;
  width: 100%;
  display: flex;
  flex-direction: column;
  align-items: center;
  p: 0;
`;

export const Content = styled.div`
  height: 100%;
  width: 100%;
  max-width: 80%;
  display: flex;
  flex-direction: column;
  align-items: center;
`;
