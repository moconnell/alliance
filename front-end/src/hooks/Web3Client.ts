import Web3Modal from "web3modal";
import WalletConnectProvider from "@walletconnect/web3-provider";
import { useEffect, useReducer, useCallback, useMemo } from "react";
import { ethers } from "ethers";
import { Web3ProviderState, web3InitialState, web3Reducer } from "../reducers";

const getWeb3Modal = () => {
  if (typeof window === "undefined") return null;

  const {
    REACT_APP_WEB3_CACHE_PROVIDER,
    REACT_APP_WEB3_INFURA_ID,
    REACT_APP_WEB3_NETWORK,
  } = process.env;

  return new Web3Modal({
    network: REACT_APP_WEB3_NETWORK,
    cacheProvider: REACT_APP_WEB3_CACHE_PROVIDER,
    providerOptions: {
      walletconnect: {
        package: WalletConnectProvider,
        options: {
          infuraId: REACT_APP_WEB3_INFURA_ID,
        },
      },
    },
  });
};

export type Web3Client = Web3ProviderState & {
  connect: () => Promise<void>;
  disconnect: () => Promise<void>;
};

export const useWeb3 = () => {
  const [state, dispatch] = useReducer(web3Reducer, web3InitialState);
  const { provider, web3Provider, address, network, signer } = state;
  const web3Modal = useMemo(() => getWeb3Modal(), []);

  const connect = useCallback(async () => {
    try {
      if (!web3Modal) {
        console.error("Cannot connect: no Web3Modal");
        return;
      }

      const provider = await web3Modal.connect();
      const web3Provider = new ethers.providers.Web3Provider(provider);
      const signer = web3Provider.getSigner();
      const address = await signer.getAddress();
      const network = await web3Provider.getNetwork();

      dispatch({
        type: "SET_WEB3_PROVIDER",
        provider,
        web3Provider,
        address,
        network,
        signer,
      });
    } catch (e) {
      console.log("connect error", e);
    }
  }, [web3Modal]);

  const disconnect = useCallback(async () => {
    if (web3Modal) {
      web3Modal.clearCachedProvider();
      if (provider?.disconnect && typeof provider.disconnect === "function") {
        await provider.disconnect();
      }
      dispatch({
        type: "RESET_WEB3_PROVIDER",
      });
    } else {
      console.error("No Web3Modal");
    }
  }, [provider, web3Modal]);

  // Auto connect to the cached provider
  useEffect(() => {
    if (web3Modal && web3Modal.cachedProvider) {
      connect();
    }
  }, [connect, web3Modal]);

  // EIP-1193 events
  useEffect(() => {
    if (provider?.on) {
      const handleAccountsChanged = (accounts: string[]) => {
        dispatch({
          type: "SET_ADDRESS",
          address: accounts[0],
        });
      };

      // https://docs.ethers.io/v5/concepts/best-practices/#best-practices--network-changes
      const handleChainChanged = (_hexChainId: string) => {
        if (typeof window !== "undefined") {
          console.log("switched to chain...", _hexChainId);
          window.location.reload();
        } else {
          console.log("window is undefined");
        }
      };

      const handleDisconnect = (error: { code: number; message: string }) => {
        // eslint-disable-next-line no-console
        console.log("disconnect", error);
        disconnect();
      };

      provider.on("accountsChanged", handleAccountsChanged);
      provider.on("chainChanged", handleChainChanged);
      provider.on("disconnect", handleDisconnect);

      // Subscription Cleanup
      return () => {
        if (provider.removeListener) {
          provider.removeListener("accountsChanged", handleAccountsChanged);
          provider.removeListener("chainChanged", handleChainChanged);
          provider.removeListener("disconnect", handleDisconnect);
        }
      };
    }
  }, [provider, disconnect]);

  return {
    provider,
    web3Provider,
    address,
    network,
    signer,
    connect,
    disconnect,
  } as Web3Client;
};
