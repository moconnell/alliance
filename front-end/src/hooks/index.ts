export type { Web3Client } from "./Web3Client";
export { useWeb3 } from "./Web3Client";
export { useCalendar } from "./useCalendar";
