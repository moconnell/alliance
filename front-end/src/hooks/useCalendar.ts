import { useCallback, useEffect, useState } from "react";

import { useWeb3Context } from "../context/Web3Context";
import assert from "assert";
import {
  Calendar,
  CalendarFactory,
  CalendarFactory__factory,
  Calendar__factory,
} from "../../typechain-types";
import DaysOfWeek from "../types/daysOfWeek";
import {
  assertValid,
  compare,
  fromTotalMins,
  totalMinutes,
} from "../utils/timeUtils";
import Meeting from "../types/meeting";
import ProfileInfo from "../types/profileInfo";
import AvailabilityInfo from "../types/availabilityInfo";
import { DateTime } from "luxon";

const NULL_ADDRESS = "0x0000000000000000000000000000000000000000";

const toString = (value: any) => (value ? String(value) : "");

export const useCalendar = () => {
  const { address, signer, web3Provider } = useWeb3Context();
  const [calendarFactory, setCalendarFactory] = useState<CalendarFactory>();
  const [calendar, setCalendar] = useState<Calendar>();
  const [availability, setAvailability] = useState<AvailabilityInfo>();
  const [profile, setProfile] = useState<ProfileInfo>();

  const { REACT_APP_WEB3_CONTRACT_FACTORY_ADDRESS } = process.env;
  assert(
    REACT_APP_WEB3_CONTRACT_FACTORY_ADDRESS,
    "REACT_APP_WEB3_CONTRACT_FACTORY_ADDRESS not set"
  );

  useEffect(() => {
    function getCalendarFactory() {
      if (!web3Provider || !address || !signer) return undefined;
      return CalendarFactory__factory.connect(
        REACT_APP_WEB3_CONTRACT_FACTORY_ADDRESS!,
        signer
      );
    }

    const fac = getCalendarFactory();
    if (fac) setCalendarFactory(fac);
  }, [REACT_APP_WEB3_CONTRACT_FACTORY_ADDRESS, address, signer, web3Provider]);

  useEffect(() => {
    async function getCalendar() {
      if (!address || !calendarFactory || !web3Provider) return;
      console.log(calendarFactory);
      const calendarAddress = await calendarFactory.userToCalendar(address);
      if (calendarAddress === NULL_ADDRESS) {
        console.log(`wallet ${address} has no saved calendar`);
        return;
      }

      console.log(`fetching calendar at ${calendarAddress}`);

      const cal = Calendar__factory.connect(calendarAddress, web3Provider);
      setCalendar(cal);
    }

    getCalendar();
  }, [address, calendarFactory, web3Provider]);

  useEffect(() => {
    async function loadProfileAvailability() {
      if (!calendar) return;

      const newProfile = await calendar.profile();
      setProfile(newProfile);

      let newAvailability = await calendar.availability();
      const from = fromTotalMins(newAvailability.earliestStartMinutes);
      const to = fromTotalMins(
        newAvailability.earliestStartMinutes + newAvailability.minutesAvailable
      );
      setAvailability({ ...newAvailability, from, to });
    }

    loadProfileAvailability();
  }, [calendar]);

  useEffect(() => {
    const createCalendar = async () => {
      if (
        calendar ||
        !calendarFactory ||
        !web3Provider ||
        !profile ||
        !availability
      )
        return;

      assert(profile.email, "No email");
      assertValid(availability.from);
      assertValid(availability.to);
      assert(
        compare(availability.from!, availability.to!) < 0,
        "from after to"
      );

      const p = {
        description: toString(profile.description),
        email: profile.email!,
        picture: toString(profile.picture),
        url: toString(profile.url),
        username: toString(profile.username),
      };

      const a = {
        availableDays: availability.availableDays ?? DaysOfWeek.None,
        earliestStartMinutes: totalMinutes(availability.from!),
        minutesAvailable:
          totalMinutes(availability.to!) - totalMinutes(availability.from!),
        location: availability.location!,
        timeZone: availability.timeZone!,
      };

      console.log("creating calendar now ... ", p, a);

      const tx = await calendarFactory.createCalendar(p, a);
      const receipt = await tx.wait();
      const txEvent = receipt.events?.[0]?.args;
      const calendarAddr = txEvent?.calendar;
      const cal = Calendar__factory.connect(calendarAddr, web3Provider);

      console.log("calendar created ok", cal);

      setCalendar(cal);
    };

    createCalendar();
  }, [
    profile,
    availability,
    calendar,
    calendarFactory,
    web3Provider,
    address,
    REACT_APP_WEB3_CONTRACT_FACTORY_ADDRESS,
  ]);

  const getMeetings = useCallback(
    async (date: Date) => {
      if (!calendar) return undefined;

      console.log(`getMeetings(date: ${date})`);

      const year = date.getFullYear();
      const month = date.getMonth() + 1;
      const day = date.getDate();
      const meetingsOnDate = await calendar.getMeetings(year, month, day);

      return meetingsOnDate.map(
        ({ attendee, startMinutes, durationMinutes }) => {
          const hour = Math.floor(startMinutes / 60);
          const minute = startMinutes % 60;
          const from = DateTime.fromObject(
            {
              year,
              month,
              day,
              hour,
              minute,
            },
            {
              zone: availability?.timeZone,
            }
          ).setZone();
          const to = from.plus({ minutes: durationMinutes });

          return {
            date: from,
            endDate: to,
            attendee: attendee,
            description: `${durationMinutes} minute meeting`,
          } as Meeting;
        }
      );
    },
    [availability?.timeZone, calendar]
  );

  const getProfileAvailability = useCallback(
    async (calendarAddress: string) => {
      if (!web3Provider) return undefined;

      console.log(
        `getProfileAvailability(calendarAddress: ${calendarAddress})`
      );

      const otherCalendar = Calendar__factory.connect(
        calendarAddress,
        web3Provider
      );

      const profileAvailability: [ProfileInfo, AvailabilityInfo] = [
        await otherCalendar.profile(),
        await otherCalendar.availability(),
      ];

      return profileAvailability;
    },
    [web3Provider]
  );

  const getAvailableTimes = useCallback(
    async (calendarAddress: string, date: Date, durationMinutes: number) => {
      if (!web3Provider) return undefined;

      console.log(
        `getAvailableTimes(calendarAddress: ${calendarAddress}, date: ${date}, duration: ${durationMinutes})`
      );

      const otherCalendar = Calendar__factory.connect(
        calendarAddress,
        web3Provider
      );

      const year = date.getFullYear();
      const month = date.getMonth() + 1;
      const day = date.getDate();
      const { times, timeZone } = await otherCalendar.getAvailableTimes(
        year,
        month,
        day,
        durationMinutes
      );

      return times
        .filter((startMinutes) => startMinutes > 0)
        .map((startMinutes) => {
          const hour = Math.floor(startMinutes / 60),
            minute = startMinutes % 60;

          return DateTime.fromObject(
            { year, month, day, hour, minute },
            { zone: timeZone }
          );
        });
    },
    [web3Provider]
  );

  const bookMeeting = useCallback(
    async (calendarAddress: string, at: DateTime, durationMinutes: number) => {
      if (!signer) return;

      console.log(
        `bookMeeting(calendarAddress: ${calendarAddress}, at: ${at}, duration: ${durationMinutes})`
      );

      const otherCalendar = Calendar__factory.connect(calendarAddress, signer);

      await otherCalendar.bookMeeting(
        at.year,
        at.month,
        at.day,
        at.hour,
        at.minute,
        durationMinutes
      );
    },
    [signer]
  );

  const setProfileAvailability = (info: ProfileInfo & AvailabilityInfo) => {
    setProfile(info);
    setAvailability(info);
  };

  return {
    address,
    availability,
    bookMeeting,
    getAvailableTimes,
    getMeetings,
    getProfileAvailability,
    hasCalendar: calendar !== undefined,
    profile,
    setProfileAvailability,
  };
};
