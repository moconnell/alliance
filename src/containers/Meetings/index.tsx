// import { Navigate } from "react-router-dom";
import MeetingCardList from "../../components/MeetingCardList";
import { useCalendar } from "../../hooks";
import Meeting from "../../types/meeting";
import { useEffect, useState } from "react";
import Calendar from "../../components/Calendar";
import { Container, Content } from "./styles";
import { Flex } from "@chakra-ui/react";

const Meetings = () => {
  const { getMeetings } = useCalendar();
  const [meetings, setMeetings] = useState([] as Meeting[]);
  const [date, setDate] = useState(new Date());

  useEffect(() => {
    async function loadMeetings() {
      const meetingsOnDate = await getMeetings(date);

      if (meetingsOnDate) {
        console.log(`Meetings on ${date}`, meetingsOnDate);
        setMeetings(meetingsOnDate);
      }
    }

    loadMeetings();
  }, [date, getMeetings]);

  // if (!hasCalendar) return <Navigate to="/" />;

  return (
    <Container>
      <Content>
        <Flex
          h={{ base: "auto", md: "100%" }}
          py={10}
          direction={{ base: "column", md: "row" }}
        >
          <Calendar defaultValue={date} onChange={setDate} />
          <MeetingCardList meetings={meetings} />
        </Flex>
      </Content>
    </Container>
  );
};

export default Meetings;
