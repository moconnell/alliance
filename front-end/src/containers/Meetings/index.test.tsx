import { render, waitFor } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";
import Meetings from "./index";
import { useCalendar } from "../../hooks";
import { DateTime } from "luxon";
import Meeting from "src/types/meeting";
import { expectsResolvedDragConstraints } from "framer-motion/types/gestures/drag/VisualElementDragControls";

jest.mock("../../hooks");
jest.mock("../../components/Calendar", () => () => "((Calendar))");

describe("<Meetings />", () => {
  const mockGetMeetings = jest.fn();

  beforeEach(() => mockGetMeetings.mockReset());
  afterAll(() => mockGetMeetings.mockRestore());

  const testData: [string, Meeting[]?][] = [
    ["render no meetings correctly"],
    [
      "render meetings correctly",
      [
        {
          date: DateTime.fromISO("2022-02-02T20:00:00").setLocale("en-gb"),
          endDate: DateTime.fromISO("2022-02-02T20:30:00").setLocale("en-gb"),
          attendee: "0xf39Fd6e51aad88F6F4ce6aB8827279cffFb92266",
          description: "30 minute meeting",
        },
      ],
    ],
  ];

  testData.forEach(([name, meetings]) =>
    it(`should ${name}`, async () => {
      (useCalendar as jest.Mock).mockImplementation(() => {
        return {
          calendar: {},
          getMeetings: mockGetMeetings,
        };
      });

      mockGetMeetings.mockImplementation(() => meetings);

      const { asFragment, getByText } = render(
        <MemoryRouter>
          <Meetings />
        </MemoryRouter>
      );

      await waitFor(() =>
        expect(getByText(meetings?.at(0)?.attendee ?? "Times shown in local time")).toBeInTheDocument()
      );

      expect(asFragment()).toMatchSnapshot();
    })
  );
});

jest.clearAllMocks();
