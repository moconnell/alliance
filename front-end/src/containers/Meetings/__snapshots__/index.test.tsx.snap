// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Meetings /> should render meetings correctly 1`] = `
<DocumentFragment>
  <div
    class="sc-iAEyYk cPjprN"
  >
    <div
      class="sc-fsQiph kiFeGQ"
    >
      <div
        class="css-b82b48"
      >
        ((Calendar))
        <div
          class="sc-jTrPJq bOCSzy"
        >
          <div
            class="sc-eDDNvR gtHwmp"
          >
            <div
              class="sc-beqWaB ljmzQY"
            >
              <span
                class="sc-hLseeU RSRGY"
              >
                20:00-20:30
              </span>
              <span
                class="sc-gueYoa kfKkec"
              >
                0xf39Fd6e51aad88F6F4ce6aB8827279cffFb92266
              </span>
              <span
                class="sc-dmqHEX gIMSGZ"
              >
                30 minute meeting
              </span>
            </div>
          </div>
          <div
            class="sc-gLDzan bxuKqx"
          >
            Times shown in local time
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`<Meetings /> should render no meetings correctly 1`] = `
<DocumentFragment>
  <div
    class="sc-iAEyYk cPjprN"
  >
    <div
      class="sc-fsQiph kiFeGQ"
    >
      <div
        class="css-b82b48"
      >
        ((Calendar))
        <div
          class="sc-jTrPJq bOCSzy"
        >
          <div
            class="sc-eDDNvR gtHwmp"
          />
          <div
            class="sc-gLDzan bxuKqx"
          >
            Times shown in local time
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
